lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1149,8 +1154,62 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         // module's ports, and use them for instance port wires. This way,
         // constraints imposed onto the ports of the instance will transparently
         // apply to the ports of the instantiated module.
-        for (auto it : llvm::zip(module.getArguments(), op->getResults())) {
-          unifyTypes(FieldRef(std::get<0>(it), 0), std::get<1>(it));
+        for (auto it : llvm::zip(op->getResults(), module.getArguments())) {
+          unifyTypes(FieldRef(std::get<0>(it), 0), FieldRef(std::get<1>(it), 0),
+                     std::get<0>(it).getType().template cast<FIRRTLType>());
+        }
+      })
+
+      // Handle memories.
+      .Case<MemOp>([&](auto op) {
+        // Create constraint variables for all ports.
+        for (auto result : op.results())
+          declareVars(result, op.getLoc());
+
+        // This creates independent variables for every data port. Yet, what we
+        // actually want is for all data ports to share the same variable. To do
+        // this, we find the first data port declared, and use that port's vars
+        // for all the other ports.
+        unsigned firstFieldID;
+        switch (op.getPortKind(0)) {
+        case MemOp::PortKind::Read:
+          firstFieldID = 3;
+          break;
+        case MemOp::PortKind::Write:
+          firstFieldID = 3;
+          break;
+        case MemOp::PortKind::ReadWrite:
+          firstFieldID = 4;
+          break;
+        }
+        FieldRef firstData(op.getResult(0), op.getPortType(0)
+                                                .getPassiveType()
+                                                .template cast<BundleType>()
+                                                .getFieldID(firstFieldID));
+        LLVM_DEBUG(llvm::dbgs() << "Adjusting memory port variables:\n");
+
+        // Reuse data port variables.
+        auto dataType = op.getDataType();
+        for (unsigned i = 0, e = op.results().size(); i < e; ++i) {
+          auto result = op.getResult(i);
+          auto portType =
+              op.getPortType(i).getPassiveType().template cast<BundleType>();
+          switch (op.getPortKind(i)) {
+          case MemOp::PortKind::Read:
+            unifyTypes(FieldRef(result, portType.getFieldID(3)), firstData,
+                       dataType);
+            break;
+          case MemOp::PortKind::Write:
+            unifyTypes(FieldRef(result, portType.getFieldID(3)), firstData,
+                       dataType);
+            break;
```

Comment:
Similar to above, you can use fallthrough to combine these cases. There's probably some further non-required refactoring to unify the logic of going from `MemOp::PortKind => Unsigned` that could be reused here. Only something to think about.

Reply:
Excellent suggestion! I added a small helper that does exactly what you say (`PortKind => ArrayRef<Unsigned>`), and it cleans things up quite a bit. Decided to stick with the switch statement for a bit of clarity and so we get a compiler warning if we ever extend the `PortKind` enum.